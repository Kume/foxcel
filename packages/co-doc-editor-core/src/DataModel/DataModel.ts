import {
  dataPathComponentIsIndexOrKey,
  dataPathComponentIsListIndex,
  dataPathComponentIsListIndexLike,
  dataPathComponentIsMapKey,
  dataPathFirstComponent,
  dataPathLength,
  ForwardDataPath,
  ForwardDataPathComponent,
  getListIndexFromDataPathElement,
  getListIndexFromDataPathElementOrFail,
  MultiDataPath,
  shiftDataPath,
} from './DataPath';
import {
  DataCollectionItem,
  DataModel,
  DataModelType,
  ListDataModel,
  MapDataModel,
  StringDataModel,
} from './DataModelTypes';
import {DataSchemaContext} from './DataSchema';
import {DataModelOperationError} from './errors';
import {ConditionConfig} from '..';

export function dataModelTypeToLabel(type: DataModelType): string {
  return DataModelType[type];
}

export function dataModelType(model: DataModel): DataModelType {
  switch (typeof model) {
    case 'number':
      if (Number.isInteger(model)) {
        return DataModelType.Integer;
      } else {
        return DataModelType.Float;
      }
    case 'boolean':
      return DataModelType.Boolean;
    case 'string':
      return DataModelType.String;
    case 'object':
      if (model === null) {
        return DataModelType.Null;
      }
      if (Array.isArray(model)) {
        return DataModelType.List;
      }
      return DataModelType.Map;
  }
}

export function unknownToDataModel(value: unknown): DataModel {
  if (value === null || value === undefined) {
    return null;
  }
  switch (typeof value) {
    case 'number':
    case 'boolean':
    case 'string':
      return value;
    case 'object':
      if (Array.isArray(value)) {
        return value.map(unknownToDataModel);
      } else {
        return {
          t: DataModelType.Map,
          v: Object.keys(value!)
            .filter((k) => (value as any)[k] !== undefined)
            .map((k) => [k, unknownToDataModel((value as any)[k])]),
        };
      }
    default:
      throw new Error('Invalid data type');
  }
}

function dataModelEquals(a: DataModel, b: DataModel): boolean {
  if (a === b) {
    return true;
  }
  if (typeof a !== 'object' || typeof b !== 'object' || a === null || b === null) {
    return false;
  }
  if (Array.isArray(a)) {
    if (!Array.isArray(b)) {
      return false;
    }
    if (a.length !== b.length) {
      return false;
    }
    return a.every((aValue, index) => dataModelEquals(aValue, b[index]));
  } else {
    if (Array.isArray(b)) {
      return false;
    }
    if (a.v.length !== b.v.length) {
      return false;
    }
    return a.v.every((_, i) => a.v[i][0] === b.v[i][0] && dataModelEquals(a.v[i][1], b.v[i][1]));
  }
}

export function dataModelEqualsToUnknown(model: DataModel, value: unknown): boolean {
  if (typeof model !== 'object' || typeof value !== 'object' || model === null || value === null) {
    return model === value;
  }
  if (Array.isArray(model)) {
    if (!Array.isArray(value)) {
      return false;
    }
    if (model.length !== value.length) {
      return false;
    }
    return model.every((modelChild, index) => dataModelEqualsToUnknown(modelChild, value[index]));
  }
  if (Array.isArray(value)) {
    return false;
  }
  const valueKeys = Object.keys(value);
  const map = mapDataModelToJsMap(model);
  if (valueKeys.length !== map.size) {
    return false;
  }
  return valueKeys.every((key) => dataModelEqualsToUnknown(map.get(key)!, (value as any)[key]));
}

export function dataModelToJson(model: DataModel): any {
  if (typeof model !== 'object' || model === null) {
    return model;
  }
  if (Array.isArray(model)) {
    return model.map(dataModelToJson);
  }
  const map: {[key: string]: any} = {};
  model.v.forEach(([key, value]) => {
    if (key !== null && map[key] === undefined) {
      map[key] = dataModelToJson(value);
    }
  });
  return map;
}

export function setToDataModel(
  path: ForwardDataPath,
  value: DataModel,
  to: DataModel,
  schema?: DataSchemaContext,
): DataModel | undefined {
  if (dataPathLength(path) === 0) {
    return dataModelEquals(value, to) ? undefined : value;
  }
  if (!dataModelIsMapOrList(to)) {
    throw new DataModelOperationError(`Cannot set data to ${dataModelTypeToLabel(dataModelType(to))}`);
  }
  if (Array.isArray(to)) {
    return setToListDataItr(to, path, schema, (nextPath, childData, childSchema) =>
      setToDataModel(nextPath, value, childData, childSchema),
    );
  }
  return setToMapDataItr(
    to,
    path,
    schema,
    (nextPath, childData, childSchema) => setToDataModel(nextPath, value, childData, childSchema),
    (key) => {
      if (dataPathLength(path) === 1 && typeof key === 'string') {
        return forceAddToMapData(to, value, key);
      }
      throw new DataModelOperationError('Cannot set data to empty value');
    },
  );
}

export function pushToDataModel(
  path: ForwardDataPath,
  value: DataModel,
  to: DataModel,
  schema?: DataSchemaContext,
): DataModel | undefined {
  const pathLength = dataPathLength(path);

  if (!dataModelIsMapOrList(to)) {
    throw new DataModelOperationError(`Cannot push data to ${dataModelTypeToLabel(dataModelType(to))}`);
  }
  if (Array.isArray(to)) {
    if (pathLength === 0) {
      return pushToListData(to, value);
    } else {
      return setToListDataItr(to, path, schema, (nextPath, childData, childSchema) =>
        pushToDataModel(nextPath, value, childData, childSchema),
      );
    }
  }
  if (pathLength === 0) {
    return pushToMapData(to, value);
  } else {
    return setToMapDataItr(
      to,
      path,
      schema,
      (nextPath, childData, childSchema) => pushToDataModel(nextPath, value, childData, childSchema),
      () => {
        throw new DataModelOperationError('Cannot push data to empty value');
      },
    );
  }
}

export function getFromDataModelForPathComponent(
  model: DataModel | undefined,
  pathComponent: ForwardDataPathComponent,
): DataModel | undefined {
  if (!model) {
    return undefined;
  }
  if (!dataModelIsMapOrList(model)) {
    return undefined;
  }
  if (Array.isArray(model)) {
    if (!dataPathComponentIsListIndexLike(pathComponent)) {
      return undefined;
    }
    const index = getListIndexFromDataPathElement(pathComponent);
    return getListDataAt(model, index);
  }
  const {index} = getMapKeyAndIndex(model, pathComponent);
  if (index === undefined) {
    return undefined;
  }
  return model.v[index][1];
}

export function getFromDataModel(model: DataModel, path: ForwardDataPath): DataModel | undefined {
  if (dataPathLength(path) === 0) {
    return model;
  }
  const firstPathComponent = dataPathFirstComponent(path);
  const childModel = getFromDataModelForPathComponent(model, firstPathComponent);
  return childModel && getFromDataModel(childModel, shiftDataPath(path));
}

type CreateNextChildData = (
  nextPath: ForwardDataPath,
  childData: DataModel,
  childSchema: DataSchemaContext | undefined,
) => DataModel | undefined;

//#region For ListDataModel
export function getListDataAt(list: ListDataModel, at: number): DataModel | undefined {
  return list[at];
}

export function setToListDataAt(list: ListDataModel, value: DataModel, at: number): ListDataModel {
  const childData = getListDataAt(list, at);
  if (!childData) {
    throw new DataModelOperationError('Cannot set data to out of index range of list.');
  }
  return dataModelEquals(childData, value) ? list : forceSetToListData(list, value, at);
}

function forceSetToListData(list: ListDataModel, value: DataModel, index: number): ListDataModel {
  const v = [...list];
  v[index] = value;
  return v;
}

function pushToListData(list: ListDataModel, value: DataModel): ListDataModel {
  return [...list, value];
}

function setToListDataItr(
  list: ListDataModel,
  path: ForwardDataPath,
  schema: DataSchemaContext | undefined,
  createNextChildData: CreateNextChildData,
): ListDataModel | undefined {
  const index = getListIndexFromDataPathElementOrFail(dataPathFirstComponent(path));
  const childData = getListDataAt(list, index);
  const childSchema = schema && schema.getListChild();
  if (!childData) {
    throw new DataModelOperationError('Cannot set data to out of index range of list.');
  }
  const nextChildData = createNextChildData(shiftDataPath(path), childData, childSchema);
  return nextChildData === undefined ? undefined : forceSetToListData(list, nextChildData, index);
}

//#endregion For ListDataModel

//#region For MapDataModel
export function findMapDataIndexOfKey(map: MapDataModel, key: string): number | undefined {
  for (let i = 0; i < map.v.length; i++) {
    if (map.v[i][0] === key) {
      return i;
    }
  }
}

export function getMapDataAt(map: MapDataModel, key: string): DataModel | undefined {
  const index = findMapDataIndexOfKey(map, key);
  return index === undefined ? undefined : map.v[index][1];
}

export function getMapDataAtIndex(map: MapDataModel, index: number): DataModel | undefined {
  return index === undefined ? undefined : map.v[index][1];
}

export function getMapKeyAtIndex(map: MapDataModel, index: number): string | null | undefined {
  return index === undefined ? undefined : map.v[index][0];
}

export function getMapKeyValueAtIndex(
  map: MapDataModel,
  index: number,
): readonly [string | null, DataModel] | undefined {
  return index === undefined ? undefined : map.v[index];
}

function forceSetToMapDataForIndex(map: MapDataModel, value: DataModel, index: number, key?: string): MapDataModel {
  const v = [...map.v];
  const newKey = key === undefined ? v[index][0] : key;
  v[index] = [newKey, value];
  return {t: DataModelType.Map, v};
}

function forceAddToMapData(map: MapDataModel, value: DataModel, key: string): MapDataModel {
  return {t: DataModelType.Map, v: [...map.v, [key, value]]};
}

function pushToMapData(map: MapDataModel, value: DataModel): MapDataModel {
  return {t: DataModelType.Map, v: [...map.v, [null, value]]};
}

function setToMapDataItr(
  map: MapDataModel,
  path: ForwardDataPath,
  schema: DataSchemaContext | undefined,
  createNextChildData: CreateNextChildData,
  onKeyMissing: (key: string | undefined) => MapDataModel | undefined,
): MapDataModel | undefined {
  const {key, index} = getMapKeyAndIndex(map, dataPathFirstComponent(path));
  if (index !== undefined) {
    const childData = map.v[index][1];
    const childSchema = schema && schema.getListChild();
    const nextChildData = createNextChildData(shiftDataPath(path), childData, childSchema);
    return nextChildData === undefined ? undefined : forceSetToMapDataForIndex(map, nextChildData, index, key);
  } else {
    return onKeyMissing(key);
  }
}

function getMapKeyAndIndex(map: MapDataModel, pathComponent: ForwardDataPathComponent): {index?: number; key?: string} {
  if (dataPathComponentIsMapKey(pathComponent)) {
    return {index: findMapDataIndexOfKey(map, pathComponent), key: pathComponent};
  } else if (dataPathComponentIsListIndex(pathComponent)) {
    const index = pathComponent >= mapDataSize(map) ? undefined : pathComponent;
    return {index, key: undefined};
  } else if (dataPathComponentIsIndexOrKey(pathComponent)) {
    const key = pathComponent.v.toString();
    return {index: findMapDataIndexOfKey(map, key), key};
  }
  throw new DataModelOperationError('Invalid path component');
}

export function mapDataSize(map: MapDataModel): number {
  return map.v.length;
}

export function mapDataModelKeys(map: MapDataModel): string[] {
  return [...new Set(mapDataModelRawValuesWithoutNullKey(map).map((i) => i[0]))];
}

export function mapDataModelToJsMap(map: MapDataModel): Map<string, DataModel> {
  return new Map(mapDataModelRawValuesWithoutNullKey(map).reverse());
}

export function mapDataModelRawValuesWithoutNullKey(map: MapDataModel): Array<[string, DataModel]> {
  return map.v.filter((i) => i[0] !== null) as any;
}

export function mapDataFirstElement(map: MapDataModel): DataModel | undefined {
  return mapDataSize(map) > 0 ? map.v[0][1] : undefined;
}

export function mapDataLastElement(map: MapDataModel): DataModel | undefined {
  return mapDataSize(map) > 0 ? map.v[map.v.length - 1][1] : undefined;
}

export function mapDataModelMapWithEmptyKey<T>(
  model: MapDataModel,
  mapper: (value: DataModel, key: string, index: number) => T,
): T[] {
  const appearedKeys = new Set<string>();
  const results: T[] = [];
  model.v.forEach(([key, value], index) => {
    if (key !== null && !appearedKeys.has(key)) {
      appearedKeys.add(key);
      mapper(value, key, index);
    }
  });
  return results;
}

export function mapDataModelMap<T>(
  model: MapDataModel,
  mapper: (value: DataModel, key: string | null, index: number) => T,
): T[] {
  return model.v.map(([key, value], index) => mapper(value, key, index));
}

//#endregion For MapDataModel

//#region For CollectionDataModel
type KeyForCollection<Model, WithEmptyKey> = Model extends MapDataModel
  ? WithEmptyKey extends true
    ? string | null
    : string
  : Model extends ListDataModel
  ? number
  : never;
type IndexForMapFunc<Model> = Model extends ListDataModel ? number : undefined;

export function dataModelMap<T, Model extends DataModel, WithEmptyKey extends boolean = false>(
  model: Model,
  mapper: (value: DataModel, key: KeyForCollection<Model, WithEmptyKey>, index: IndexForMapFunc<Model>) => T,
  withEmptyKey?: WithEmptyKey,
): T[] {
  if (!dataModelIsMapOrList(model)) {
    return [];
  }

  if (Array.isArray(model)) {
    return model.map(mapper as any);
  } else {
    return withEmptyKey ? mapDataModelMapWithEmptyKey(model, mapper as any) : mapDataModelMap(model, mapper as any);
  }
}

//#endregion For CollectionDataModel

export function dataModelIsMapOrList(model: DataModel): model is MapDataModel | ListDataModel {
  return typeof model === 'object' && model !== null;
}

export function dataModelIsMap(model: DataModel): model is MapDataModel {
  return typeof model === 'object' && model !== null && !Array.isArray(model);
}

export function dataModelIsInteger(model: DataModel): model is number {
  return typeof model === 'number' && Number.isInteger(model);
}

export function dataModelIsString(model: DataModel): model is StringDataModel {
  return typeof model === 'string';
}

export function dataModelIsList(model: DataModel): model is ListDataModel {
  return typeof model === 'object' && model !== null && Array.isArray(model);
}

export function stringDataModelToString(model: string): string {
  return model;
}

export function numberDataModelToNumber(model: number): number {
  return model;
}

export function conditionIsMatch(
  condition: ConditionConfig<MultiDataPath>,
  collectDataForPath: undefined | ((path: MultiDataPath) => DataCollectionItem[]),
): boolean {
  if ('or' in condition) {
    return condition.or.some((i) => conditionIsMatch(i, collectDataForPath));
  } else if ('and' in condition) {
    return condition.and.some((i) => conditionIsMatch(i, collectDataForPath));
  } else if ('match' in condition) {
    const targetModels = collectDataForPath?.(condition.path) || [];
    return targetModels.some(({data}) => dataModelEqualsToUnknown(data, condition.match));
  } else {
    throw new Error('Invalid condition');
  }
}
