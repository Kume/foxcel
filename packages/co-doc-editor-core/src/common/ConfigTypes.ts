export type RootSchemaConfig = {
  readonly namedDataSchema?: SubDataSchemaConfig;
  readonly dataSchema: DataSchemaConfig;
  readonly namedUiSchema?: SubUISchemaConfig;
  readonly fileMap: DataMapperConfig;
} & UISchemaConfigHolder;

type UISchemaConfigHolder = {readonly uiRoot: UISchemaConfig /* old schema */} | {readonly uiSchema: UISchemaConfig};

export type SubDataSchemaConfig = {readonly [key: string]: string | DataSchemaConfig};

export type SubUISchemaConfig = {readonly [key: string]: UISchemaConfig | string};

////////////////////////////////////////////////////////////////////////////
// DataMapper Config
////////////////////////////////////////////////////////////////////////////

export type DataMapperNodeConfig = MapDataMapperNodeConfig | SingleDataMapperNodeConfig;

interface DataMapperNodeConfigBase {
  path: string;
  children?: Array<DataMapperNodeConfig>;
}

interface MapDataMapperNodeConfig extends DataMapperNodeConfigBase {
  type: 'map';
  directory: string;
}

interface SingleDataMapperNodeConfig extends DataMapperNodeConfigBase {
  type: 'single';
  directory?: string;
  fileName: string;
}

export interface DataMapperConfig {
  children: Array<DataMapperNodeConfig>;
}

////////////////////////////////////////////////////////////////////////////
// DataSchema Config
////////////////////////////////////////////////////////////////////////////
export interface DataSchemaConfigBase {
  label?: string;
  dataLabel?: string;
  dataDescription?: string;
}

export type DataSchemaConfig =
  | NumberDataSchemaConfig
  | BooleanDataSchemaConfig
  | StringDataSchemaConfig
  | MapDataSchemaConfig
  | FixedMapDataSchemaConfig
  | ListDataSchemaConfig
  | ConditionalDataSchemaConfig;

export type DataSchemaConfigType = DataSchemaConfig['type'];

export interface NumberDataSchemaConfig extends DataSchemaConfigBase {
  type: 'number';
}

export interface BooleanDataSchemaConfig extends DataSchemaConfigBase {
  type: 'boolean';
}

export interface StringDataSchemaConfig extends DataSchemaConfigBase {
  type: 'string';
  in?: SelectOptionConfig<string>;
}

export interface MapDataSchemaConfig extends DataSchemaConfigBase {
  type: 'map';
  item: DataSchemaConfig | string;
}

export interface FixedMapDataSchemaConfig extends DataSchemaConfigBase {
  type: 'fixed_map';
  items: {[key: string]: DataSchemaConfig | string};
}

export interface ListDataSchemaConfig extends DataSchemaConfigBase {
  type: 'list';
  item: DataSchemaConfig | string;
}

export interface ConditionalDataSchemaConfig extends DataSchemaConfigBase {
  type: 'conditional';
  items: {[key: string]: ConditionalDataSchemaItemConfig};
}

////////////////////////////////////////////////////////////////////////////
// Conditional Config
////////////////////////////////////////////////////////////////////////////
export type MatchConditionConfig<Path = string, Value = unknown> = {path: Path; readonly match: Value};
export type OrConditionConfig<Path = string, Value = unknown> = {readonly or: readonly ConditionConfig<Path, Value>[]};
export type AndConditionConfig<Path = string, Value = unknown> = {readonly and: readonly ConditionConfig<Path>[]};

export type ConditionConfig<Path = string, Value = unknown> =
  | MatchConditionConfig<Path, Value>
  | AndConditionConfig<Path, Value>
  | OrConditionConfig<Path, Value>;

export interface ConditionalSchemaItem<Item, Path = string, Value = unknown> {
  condition: ConditionConfig<Path, Value>;
  item: Item;
}

export type ConditionalDataSchemaItemConfig = ConditionalSchemaItem<DataSchemaConfig | string>;

////////////////////////////////////////////////////////////////////////////
// Select Option Config
////////////////////////////////////////////////////////////////////////////
export interface SelectDynamicOptionConfig {
  path: string;
  labelPath?: string;
  valuePath?: string;
}

export interface SelectStaticOptionConfig<T = number | string> {
  label: string;
  value: T;
}

export type SelectOptionConfigItem<T = number | string> =
  | string
  | SelectStaticOptionConfig<T>
  | SelectDynamicOptionConfig;

export type SelectOptionConfig<T = number | string> = SelectDynamicOptionConfig | readonly SelectOptionConfigItem<T>[];

////////////////////////////////////////////////////////////////////////////
// UIConfig Config
////////////////////////////////////////////////////////////////////////////
interface UISchemaConfigBase {
  readonly label?: string;
  readonly key?: string;
}

export type KeyOrKeyFlatten = {key?: string; keyFlatten?: false} | {keyFlatten: true};

export type UISchemaConfig =
  | MappingTableUISchemaConfig
  | CheckBoxUISchemaConfig
  | ContentListUISchemaConfig
  | FormUISchemaConfig
  | NumberUISchemaConfig
  | SelectUISchemaConfig
  | TableUISchemaConfig
  | TabUISchemaConfig
  | TextUISchemaConfig
  | ConditionalUISchemaConfig;

export type UISchemaConfigType = UISchemaConfig['type'];

export interface MappingTableUISchemaConfig extends UISchemaConfigBase {
  readonly type: 'mappingTable';
  readonly sourcePath: string;
  readonly contents: ReadonlyArray<UISchemaConfig | string>;
}

export interface TextUISchemaConfig extends UISchemaConfigBase {
  readonly type: 'text';
  readonly multiline?: boolean;
  readonly options?: ReadonlyArray<string>;
  // readonly references?: { [key: string]: TemplateReferenceConfig };
}

export type TabUISchemaConfig = {
  readonly type: 'tab';
  readonly label?: string;
  readonly contents: ReadonlyArray<UISchemaConfig | string>;
} & KeyOrKeyFlatten;

export interface TableUISchemaConfig extends UISchemaConfigBase {
  readonly type: 'table';
  readonly dataType?: CollectionDataModelTypeString;
  readonly contents: ReadonlyArray<UISchemaConfig | string>;
}

export interface SelectUISchemaConfig extends UISchemaConfigBase {
  readonly type: 'select';
  readonly emptyToNull?: boolean;
  readonly options?: SelectOptionConfig;
  readonly isMulti?: boolean;
}

export interface NumberUISchemaConfig extends UISchemaConfigBase {
  readonly type: 'number';
}

export type FormUISchemaConfig = {
  readonly type: 'form';
  readonly label?: string;
  readonly contents: Array<UISchemaConfig | string>;
} & KeyOrKeyFlatten;

export interface ContentListUISchemaConfig extends UISchemaConfigBase {
  readonly type: 'contentList';
  readonly listIndexKey?: string;
  readonly dataType?: CollectionDataModelTypeString;
  readonly content: UISchemaConfig | string;
}

export interface CheckBoxUISchemaConfig extends UISchemaConfigBase {
  readonly type: 'checkbox';
}

export interface ConditionalUISchemaConfig extends UISchemaConfigBase {
  readonly type: 'conditional';
  readonly conditions?: {readonly [key: string]: ConditionConfig};
  readonly conditionalContents: {readonly [key: string]: UISchemaConfig | string};
}

export type CollectionDataModelTypeString = 'list' | 'map';

// export interface TemplateReferencePathConfig {
//   readonly path: string;
//   readonly keyPath: string;
//   readonly description?: string;
// }
//
// export interface TemplateReferenceConfig {
//   readonly name?: string;
//   readonly paths: (TemplateReferencePathConfig | TemplateReferencePathConfig[])[];
// }
