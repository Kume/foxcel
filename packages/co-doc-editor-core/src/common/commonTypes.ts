export interface NamedItemNode<T> {
  children?: Map<string, NamedItemNode<T>>;
  named?: Map<string, T>;
  filePath: string[];
}

export interface CommonReferenceSchema {
  readonly ref: string;
  readonly namespaceRef?: string;
  readonly namespace: readonly string[];
}
