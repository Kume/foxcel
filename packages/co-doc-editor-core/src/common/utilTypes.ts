export type Writable<T> = {-readonly [K in keyof T]: Writable<T[K]>};
export type ShallowWritable<T> = {-readonly [K in keyof T]: T[K]};
