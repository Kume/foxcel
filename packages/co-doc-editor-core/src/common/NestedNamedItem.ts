import {NamedItemNode} from './commonTypes';

export class NestedNamedItem<T> {
  public constructor(public readonly root: NamedItemNode<T>) {}

  public resolve(ref: string, namespaceRef: string | undefined, namespace: readonly string[]): T | undefined {
    let node = this.getNodeForNamespace(namespace);
    if (namespaceRef) {
      namespace = [...namespace, namespaceRef];
      node = this.getChildNode(node, namespaceRef, namespace);
    }
    return this.getNamedItem(node, ref, namespace);
  }

  private getNamedItem(node: NamedItemNode<T>, name: string, namespace: readonly string[]): T {
    const item = node.named?.get(name);
    if (!item) {
      throw new Error(`named item not found for ${namespace.join()}`);
    }
    return item;
  }

  private getChildNode(node: NamedItemNode<T>, name: string, namespace: readonly string[]): NamedItemNode<T> {
    const childNode = node.children?.get(name);
    if (!childNode) {
      throw new Error(`named node not found for ${namespace.join()}`);
    }
    return childNode;
  }

  private getNodeForNamespace(namespace: readonly string[]): NamedItemNode<T> {
    let currentNode: NamedItemNode<T> = this.root;
    const currentNamespace: string[] = [];
    namespace.forEach((name) => {
      currentNamespace.push(name);
      currentNode = this.getChildNode(currentNode, name, currentNamespace);
    });
    return currentNode;
  }
}
