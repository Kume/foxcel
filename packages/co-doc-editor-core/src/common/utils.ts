import {NamedItemNode} from './commonTypes';

export function mapObjectToObject<V, R>(
  obj: {[key: string]: V},
  mapper: (value: V, key: string) => R,
): {readonly [key: string]: R} {
  const ret: {[key: string]: R} = {};
  for (const key of Object.keys(obj)) {
    ret[key] = mapper(obj[key], key);
  }
  return ret;
}

export function mapToObject<V, R>(
  values: readonly V[],
  mapper: (value: V, index: number) => readonly [string, R | undefined] | undefined,
  ignoreUndefined: true,
): {[key: string]: R};
export function mapToObject<V, R>(
  values: readonly V[],
  mapper: (value: V, index: number) => readonly [string, R] | undefined,
  ignoreUndefined?: false,
): {[key: string]: R};
export function mapToObject<V, R>(
  values: readonly V[],
  mapper: (value: V, index: number) => readonly [string, R] | undefined,
  ignoreUndefined?: boolean,
): {[key: string]: R} {
  const ret: {[key: string]: R} = {};
  for (let i = 0; i < values.length; i++) {
    const mapped = mapper(values[i], i);
    if (mapped) {
      const [key, value] = mapped;
      if (!ignoreUndefined || value !== undefined) {
        ret[key] = value;
      }
    }
  }
  return ret;
}

export function addChildToNamedItemNode<T>(node: NamedItemNode<T>, name: string, child: NamedItemNode<T>): void {
  if (!node.children) {
    node.children = new Map();
  }
  node.children.set(name, child);
}

export function addNamedItemToNamedItemNode<T>(node: NamedItemNode<T>, name: string, item: T): void {
  if (!node.named) {
    node.named = new Map();
  }
  node.named.set(name, item);
}

export function resolvePath(currentPath: readonly string[], targetPath: readonly string[]): string[] {
  const path = [...currentPath, ...targetPath];
  let i = 0;
  while (i < path.length) {
    if (path[i] === '..') {
      if (i === 0) {
        throw new Error('Cannot reference above the top directory');
      }
      path.splice(i - 1, 2);
      i--;
    } else {
      i++;
    }
  }
  return path;
}
