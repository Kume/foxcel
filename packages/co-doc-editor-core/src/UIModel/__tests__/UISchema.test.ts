import {
  createRootUiSchemaParsingContext,
  getUiSchemaUniqueKeyOrUndefined,
  parseUISchemaConfig2,
  TabUISchema,
  UISchema,
} from '../UISchema';
import {UISchemaConfig} from '../..';
import {DataSchema, DataSchemaType} from '../../DataModel/DataSchema';
import {mapToObject} from '../../common/utils';

const emptyRootContext = createRootUiSchemaParsingContext(undefined);

type TestData = {readonly uiSchema: UISchema; readonly dataSchema: DataSchema};
type TestDataParam = {readonly key?: string};

const testDataCreator = {
  number: (): TestData => {
    const dataSchema = {t: DataSchemaType.Number} as const;
    const uiSchema = {type: 'number', key: 'test_number', dataSchema} as const;
    return {uiSchema, dataSchema} as const;
  },
  text: (): TestData => {
    const dataSchema = {t: DataSchemaType.String} as const;
    const uiSchema = {type: 'text', key: 'test_text', dataSchema} as const;
    return {uiSchema, dataSchema} as const;
  },
  form: (param: TestDataParam, children: readonly TestData[]): TestData => {
    const dataSchema = {
      t: DataSchemaType.FixedMap,
      items: mapToObject(children, (child) => {
        const key = getUiSchemaUniqueKeyOrUndefined(child.uiSchema);
        return key ? [key, child.dataSchema] : undefined;
      }),
    } as const;
    const uiSchema = {
      type: 'form',
      key: param.key!,
      contents: children.map((child) => child.uiSchema),
      dataSchema,
    } as const;
    return {uiSchema, dataSchema};
  },
  tab: (param: TestDataParam, children: readonly TestData[]): TestData => {
    const dataSchema = {
      t: DataSchemaType.FixedMap,
      items: mapToObject(children, (child) => {
        const key = getUiSchemaUniqueKeyOrUndefined(child.uiSchema);
        return key ? [key, child.dataSchema] : undefined;
      }),
    } as const;
    const uiSchema: TabUISchema = {
      type: 'tab',
      key: param.key!,
      contents: children.map((child) => child.uiSchema),
      dataSchema,
    } as const;
    return {uiSchema, dataSchema};
  },
};

describe('Unit tests for parseUISchemaConfig', () => {
  describe('For text', () => {
    it('DataSchema is empty', () => {
      const result = parseUISchemaConfig2({type: 'text'}, new Map(), emptyRootContext, undefined);
      expect(result.type).toBe('text');
      expect(result.dataSchema).toEqual({t: DataSchemaType.String});
    });
  });

  describe('For number', () => {
    it('DataSchema is empty', () => {
      const result = parseUISchemaConfig2({type: 'number'}, new Map(), emptyRootContext, undefined);
      expect(result.type).toBe('number');
      expect(result.dataSchema).toEqual({t: DataSchemaType.Number});
    });
  });

  describe('For checkbox', () => {
    it('DataSchema is empty', () => {
      const result = parseUISchemaConfig2({type: 'checkbox'}, new Map(), emptyRootContext, undefined);
      expect(result.type).toBe('checkbox');
      expect(result.dataSchema).toEqual({t: DataSchemaType.Boolean});
    });
  });

  describe('For form', () => {
    const configWithKeyFlatten: UISchemaConfig = {
      type: 'form',
      contents: [
        {type: 'text', key: 'test_text'},
        {type: 'form', key: 'test_form', contents: [{type: 'number', key: 'test_number'}]},
      ],
    };
    describe('Without DataSchema', () => {
      it('Simple', () => {
        const uiSchemaConfig: UISchemaConfig = {type: 'form', contents: [{type: 'text', key: 'test_text'}]};
        const result = parseUISchemaConfig2(uiSchemaConfig, new Map(), emptyRootContext, undefined);
        expect(result).toEqual({
          type: 'form',
          contents: [testDataCreator.text().uiSchema],
          dataSchema: {
            t: DataSchemaType.FixedMap,
            items: {test_text: {t: DataSchemaType.String}},
          },
        });
      });

      it('Nested without keyFlatten', () => {
        const result = parseUISchemaConfig2(configWithKeyFlatten, new Map(), emptyRootContext, undefined);
        const expected = testDataCreator.form({}, [
          testDataCreator.text(),
          testDataCreator.form({key: 'test_form'}, [testDataCreator.number()]),
        ]);
        expect(result).toEqual(expected.uiSchema);
        expect(result).toMatchSnapshot();
      });
    });

    describe('With DataSchema', () => {
      it('Nested without keyFlatten', () => {
        const expected = testDataCreator.form({}, [
          testDataCreator.text(),
          testDataCreator.form({key: 'test_form'}, [testDataCreator.number()]),
        ]);
        const result = parseUISchemaConfig2(configWithKeyFlatten, new Map(), emptyRootContext, expected.dataSchema);
        expect(result).toEqual(expected.uiSchema);
        expect(result).toMatchSnapshot();
      });
    });
  });
});
