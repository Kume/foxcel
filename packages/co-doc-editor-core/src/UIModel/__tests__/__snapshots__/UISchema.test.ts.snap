// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit tests for parseUISchemaConfig For form With DataSchema Nested without keyFlatten 1`] = `
Object {
  "contents": Array [
    Object {
      "dataSchema": Object {
        "t": 2,
      },
      "key": "test_text",
      "type": "text",
    },
    Object {
      "contents": Array [
        Object {
          "dataSchema": Object {
            "t": 0,
          },
          "key": "test_number",
          "type": "number",
        },
      ],
      "dataSchema": Object {
        "items": Object {
          "test_number": Object {
            "t": 0,
          },
        },
        "t": 4,
      },
      "key": "test_form",
      "type": "form",
    },
  ],
  "dataSchema": Object {
    "items": Object {
      "test_form": Object {
        "items": Object {
          "test_number": Object {
            "t": 0,
          },
        },
        "t": 4,
      },
      "test_text": Object {
        "t": 2,
      },
    },
    "t": 4,
  },
  "key": undefined,
  "type": "form",
}
`;

exports[`Unit tests for parseUISchemaConfig For form Without DataSchema Nested without keyFlatten 1`] = `
Object {
  "contents": Array [
    Object {
      "dataSchema": Object {
        "t": 2,
      },
      "key": "test_text",
      "type": "text",
    },
    Object {
      "contents": Array [
        Object {
          "dataSchema": Object {
            "t": 0,
          },
          "key": "test_number",
          "type": "number",
        },
      ],
      "dataSchema": Object {
        "items": Object {
          "test_number": Object {
            "t": 0,
          },
        },
        "t": 4,
      },
      "key": "test_form",
      "type": "form",
    },
  ],
  "dataSchema": Object {
    "items": Object {
      "test_form": Object {
        "items": Object {
          "test_number": Object {
            "t": 0,
          },
        },
        "t": 4,
      },
      "test_text": Object {
        "t": 2,
      },
    },
    "t": 4,
  },
  "key": undefined,
  "type": "form",
}
`;
